<h1>Activiy 3</h1>
<hr>
<h2>Analyst: Reniel P. Bontoc</h2>

In [1]:
import numpy as np
import pandas as pd

In [2]:
penguins = pd.read_csv('datasets\\penguins.csv')
uk = pd.read_csv('datasets\\uk_road_accident.csv')

In [3]:
penguins

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Adelie  Torgersen            39.1           18.7              181.0   
1       Adelie  Torgersen            39.5           17.4              186.0   
2       Adelie  Torgersen            40.3           18.0              195.0   
3       Adelie  Torgersen             NaN            NaN                NaN   
4       Adelie  Torgersen            36.7           19.3              193.0   
..         ...        ...             ...            ...                ...   
339  Chinstrap      Dream            55.8           19.8              207.0   
340  Chinstrap      Dream            43.5           18.1              202.0   
341  Chinstrap      Dream            49.6           18.2              193.0   
342  Chinstrap      Dream            50.8           19.0              210.0   
343  Chinstrap      Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  
0         3750.0    male  2007  
1         3800.0  female  2007  
2         3250.0  female  2007  
3            NaN     NaN  2007  
4         3450.0  female  2007  
..           ...     ...   ...  
339       4000.0    male  2009  
340       3400.0  female  2009  
341       3775.0    male  2009  
342       4100.0    male  2009  
343       3775.0  female  2009  

[344 rows x 8 columns]

In [4]:
uk

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [5]:
uk.rename(columns={
    "Accident Date": "Accident_Date",
    "District Area": "District_Area"
}, inplace=True)

In [6]:
uk

Index Accident_Severity Accident_Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District_Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

<h1>Filling null values</h1>
<hr>

<h1>PENGUINS</h1>

In [7]:
penguins['bill_length_mm'].head()
penguins['bill_length_mm'].isnull().sum()

np.int64(2)

In [8]:
penguins['bill_length_mm'] = penguins['bill_length_mm'].fillna(penguins['bill_length_mm'].mean())
penguins.isnull().sum()

species               0
island                0
bill_length_mm        0
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [9]:
penguins['bill_depth_mm'] = penguins['bill_depth_mm'].fillna(penguins['bill_depth_mm'].mean())
penguins.isnull().sum()

species               0
island                0
bill_length_mm        0
bill_depth_mm         0
flipper_length_mm     2
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [10]:
penguins['flipper_length_mm'] = penguins['flipper_length_mm'].fillna(penguins['flipper_length_mm'].mean())
penguins.isnull().sum()

species               0
island                0
bill_length_mm        0
bill_depth_mm         0
flipper_length_mm     0
body_mass_g           2
sex                  11
year                  0
dtype: int64

In [11]:
penguins['body_mass_g'] = penguins['body_mass_g'].fillna(penguins['body_mass_g'].mean())
penguins.isnull().sum()

species               0
island                0
bill_length_mm        0
bill_depth_mm         0
flipper_length_mm     0
body_mass_g           0
sex                  11
year                  0
dtype: int64

In [12]:
penguins['sex'] = penguins['sex'].fillna(penguins['sex'].mode()[0])
penguins.isnull().sum()

species              0
island               0
bill_length_mm       0
bill_depth_mm        0
flipper_length_mm    0
body_mass_g          0
sex                  0
year                 0
dtype: int64

In [13]:
penguins.columns

Index(['species', 'island', 'bill_length_mm', 'bill_depth_mm',
       'flipper_length_mm', 'body_mass_g', 'sex', 'year'],
      dtype='object')

In [14]:
penguins['mass_kg'] = np.round(penguins['body_mass_g'] / 1000, 1)

In [15]:
penguins

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0       Adelie  Torgersen        39.10000       18.70000         181.000000   
1       Adelie  Torgersen        39.50000       17.40000         186.000000   
2       Adelie  Torgersen        40.30000       18.00000         195.000000   
3       Adelie  Torgersen        43.92193       17.15117         200.915205   
4       Adelie  Torgersen        36.70000       19.30000         193.000000   
..         ...        ...             ...            ...                ...   
339  Chinstrap      Dream        55.80000       19.80000         207.000000   
340  Chinstrap      Dream        43.50000       18.10000         202.000000   
341  Chinstrap      Dream        49.60000       18.20000         193.000000   
342  Chinstrap      Dream        50.80000       19.00000         210.000000   
343  Chinstrap      Dream        50.20000       18.70000         198.000000   

     body_mass_g     sex  year  mass_kg  
0    3750.000000    male  2007      3.8  
1    3800.000000  female  2007      3.8  
2    3250.000000  female  2007      3.2  
3    4201.754386    male  2007      4.2  
4    3450.000000  female  2007      3.4  
..           ...     ...   ...      ...  
339  4000.000000    male  2009      4.0  
340  3400.000000  female  2009      3.4  
341  3775.000000    male  2009      3.8  
342  4100.000000    male  2009      4.1  
343  3775.000000  female  2009      3.8  

[344 rows x 9 columns]

In [16]:
penguins['mass_kg'].max()

np.float64(6.3)

In [17]:
penguins[penguins['mass_kg'] == 6.3]

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
169  Gentoo  Biscoe            49.2           15.2              221.0   

     body_mass_g   sex  year  mass_kg  
169       6300.0  male  2007      6.3

<h1>UK ROAD ACCIDENTS</h1>

In [18]:
uk

Index Accident_Severity Accident_Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District_Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [19]:
uk['Latitude'] = uk['Latitude'].fillna(uk['Latitude'].mean())
uk['Longitude'] = uk['Longitude'].fillna(uk['Longitude'].mean())
uk['Road_Surface_Conditions'] = uk['Road_Surface_Conditions'].fillna(uk['Road_Surface_Conditions'].mode()[0])
uk['Road_Type'] = uk['Road_Type'].fillna(uk['Road_Type'].mode()[0])
uk['Urban_or_Rural_Area'] = uk['Urban_or_Rural_Area'].fillna(uk['Urban_or_Rural_Area'].mode()[0])
uk['Weather_Conditions'] = uk['Weather_Conditions'].fillna(uk['Weather_Conditions'].mode()[0])
uk.isnull().sum()

Index                      0
Accident_Severity          0
Accident_Date              0
Latitude                   0
Light_Conditions           0
District_Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

<h1>Questions:</h1>
<hr>

<h2>1. Which penguin is the heaviest?</h2>

In [20]:
penguins['mass_kg'].max()

np.float64(6.3)

In [21]:
penguins[penguins['mass_kg'] == 6.3]

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
169  Gentoo  Biscoe            49.2           15.2              221.0   

     body_mass_g   sex  year  mass_kg  
169       6300.0  male  2007      6.3

<h2>Insights:</h2>
<h3>The heaviest penguin in the dataset is a male Gentoo from Biscoe Island in 2007, weighing 6.3 kg. It also has one of the longer flipper lengths (221 mm) and a relatively large bill length (49.2 mm).</h3>
<hr>

<h2>2. Which penguin is the most light?</h2>

In [22]:
penguins['mass_kg'].min()

np.float64(2.7)

In [23]:
penguins[penguins['mass_kg'] == 2.7]

species island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
314  Chinstrap  Dream            46.9           16.6              192.0   

     body_mass_g     sex  year  mass_kg  
314       2700.0  female  2008      2.7

<h2>Insights:</h2>
<h3>The lightest penguin in the dataset is a female Chinstrap from Dream Island in 2008, weighing 2.7 kg. It has a smaller flipper length (192 mm) compared to the heaviest penguin and a slightly shorter bill length (46.9 mm).</h3>
<hr>

<h2>3. How many penguins in Biscoe island?</h2>

In [24]:
biscoe = penguins[penguins['island'] == 'Biscoe']
biscoe

species  island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
20   Adelie  Biscoe        37.80000       18.30000         174.000000   
21   Adelie  Biscoe        37.70000       18.70000         180.000000   
22   Adelie  Biscoe        35.90000       19.20000         189.000000   
23   Adelie  Biscoe        38.20000       18.10000         185.000000   
24   Adelie  Biscoe        38.80000       17.20000         180.000000   
..      ...     ...             ...            ...                ...   
271  Gentoo  Biscoe        43.92193       17.15117         200.915205   
272  Gentoo  Biscoe        46.80000       14.30000         215.000000   
273  Gentoo  Biscoe        50.40000       15.70000         222.000000   
274  Gentoo  Biscoe        45.20000       14.80000         212.000000   
275  Gentoo  Biscoe        49.90000       16.10000         213.000000   

     body_mass_g     sex  year  mass_kg  
20   3400.000000  female  2007      3.4  
21   3600.000000    male  2007      3.6  
22   3800.000000  female  2007      3.8  
23   3950.000000    male  2007      4.0  
24   3800.000000    male  2007      3.8  
..           ...     ...   ...      ...  
271  4201.754386    male  2009      4.2  
272  4850.000000  female  2009      4.8  
273  5750.000000    male  2009      5.8  
274  5200.000000  female  2009      5.2  
275  5400.000000    male  2009      5.4  

[168 rows x 9 columns]

<h2>Insights:</h2>
<h3>Biscoe Island has 168 penguins in the dataset, composed of Adelie and Gentoo species. The data includes both males and females, with varied physical measurements — flipper lengths range from 174 mm to 222 mm, and body mass from 3.4 kg to 5.8 kg across the recorded years.</h3>
<hr>

<h2>4. How many female penguins are in Dream Island</h2>

In [25]:
female_dream_penguins = penguins[(penguins['sex'] == 'female') & (penguins['island'] == 'Dream')]
female_dream_penguins

species island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
30      Adelie  Dream            39.5           16.7              178.0   
32      Adelie  Dream            39.5           17.8              188.0   
34      Adelie  Dream            36.4           17.0              195.0   
37      Adelie  Dream            42.2           18.5              180.0   
38      Adelie  Dream            37.6           19.3              181.0   
..         ...    ...             ...            ...                ...   
335  Chinstrap  Dream            45.6           19.4              194.0   
337  Chinstrap  Dream            46.8           16.5              189.0   
338  Chinstrap  Dream            45.7           17.0              195.0   
340  Chinstrap  Dream            43.5           18.1              202.0   
343  Chinstrap  Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  mass_kg  
30        3250.0  female  2007      3.2  
32        3300.0  female  2007      3.3  
34        3325.0  female  2007      3.3  
37        3550.0  female  2007      3.6  
38        3300.0  female  2007      3.3  
..           ...     ...   ...      ...  
335       3525.0  female  2009      3.5  
337       3650.0  female  2009      3.6  
338       3650.0  female  2009      3.6  
340       3400.0  female  2009      3.4  
343       3775.0  female  2009      3.8  

[61 rows x 9 columns]

<h2>Insights:</h2>
<h3>The data shows that there are 61 female penguins on Dream Island, consisting mainly of Adelie and Chinstrap species. Their body mass ranges from 3.25 kg to 3.78 kg, with flipper lengths between 178 mm and 202 mm, and bill lengths from 36.4 mm to 50.2 mm</h3>
<hr>

<h2>5. Penguins with flipper length greater than 200 mm</h2>

In [26]:
long_flipper_penguins = penguins[penguins['flipper_length_mm'] > 200]
long_flipper_penguins

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
3       Adelie  Torgersen        43.92193       17.15117         200.915205   
90      Adelie      Dream        35.70000       18.00000         202.000000   
91      Adelie      Dream        41.10000       18.10000         205.000000   
95      Adelie      Dream        40.80000       18.90000         208.000000   
101     Adelie     Biscoe        41.00000       20.00000         203.000000   
..         ...        ...             ...            ...                ...   
334  Chinstrap      Dream        50.20000       18.80000         202.000000   
336  Chinstrap      Dream        51.90000       19.50000         206.000000   
339  Chinstrap      Dream        55.80000       19.80000         207.000000   
340  Chinstrap      Dream        43.50000       18.10000         202.000000   
342  Chinstrap      Dream        50.80000       19.00000         210.000000   

     body_mass_g     sex  year  mass_kg  
3    4201.754386    male  2007      4.2  
90   3550.000000  female  2008      3.6  
91   4300.000000    male  2008      4.3  
95   4300.000000    male  2008      4.3  
101  4725.000000    male  2009      4.7  
..           ...     ...   ...      ...  
334  3800.000000    male  2009      3.8  
336  3950.000000    male  2009      4.0  
339  4000.000000    male  2009      4.0  
340  3400.000000  female  2009      3.4  
342  4100.000000    male  2009      4.1  

[150 rows x 9 columns]

<h2>Insihgts:</h2>
<h3>The dataset shows 150 penguins with flipper lengths greater than 200 mm, spanning species like Adelie and Chinstrap across multiple islands. These penguins generally have larger body masses, ranging from about 3.4 kg to 4.7 kg</h3>
<hr>

<h2>6. Penguins with flipper length greater than 200 mm</h2>

In [27]:
adelie_torgersen_penguins = penguins[(penguins['species'] == 'Adelie') & (penguins['island'] == 'Torgersen')]
adelie_torgersen_penguins

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen        39.10000       18.70000         181.000000   
1    Adelie  Torgersen        39.50000       17.40000         186.000000   
2    Adelie  Torgersen        40.30000       18.00000         195.000000   
3    Adelie  Torgersen        43.92193       17.15117         200.915205   
4    Adelie  Torgersen        36.70000       19.30000         193.000000   
5    Adelie  Torgersen        39.30000       20.60000         190.000000   
6    Adelie  Torgersen        38.90000       17.80000         181.000000   
7    Adelie  Torgersen        39.20000       19.60000         195.000000   
8    Adelie  Torgersen        34.10000       18.10000         193.000000   
9    Adelie  Torgersen        42.00000       20.20000         190.000000   
10   Adelie  Torgersen        37.80000       17.10000         186.000000   
11   Adelie  Torgersen        37.80000       17.30000         180.000000   
12   Adelie  Torgersen        41.10000       17.60000         182.000000   
13   Adelie  Torgersen        38.60000       21.20000         191.000000   
14   Adelie  Torgersen        34.60000       21.10000         198.000000   
15   Adelie  Torgersen        36.60000       17.80000         185.000000   
16   Adelie  Torgersen        38.70000       19.00000         195.000000   
17   Adelie  Torgersen        42.50000       20.70000         197.000000   
18   Adelie  Torgersen        34.40000       18.40000         184.000000   
19   Adelie  Torgersen        46.00000       21.50000         194.000000   
68   Adelie  Torgersen        35.90000       16.60000         190.000000   
69   Adelie  Torgersen        41.80000       19.40000         198.000000   
70   Adelie  Torgersen        33.50000       19.00000         190.000000   
71   Adelie  Torgersen        39.70000       18.40000         190.000000   
72   Adelie  Torgersen        39.60000       17.20000         196.000000   
73   Adelie  Torgersen        45.80000       18.90000         197.000000   
74   Adelie  Torgersen        35.50000       17.50000         190.000000   
75   Adelie  Torgersen        42.80000       18.50000         195.000000   
76   Adelie  Torgersen        40.90000       16.80000         191.000000   
77   Adelie  Torgersen        37.20000       19.40000         184.000000   
78   Adelie  Torgersen        36.20000       16.10000         187.000000   
79   Adelie  Torgersen        42.10000       19.10000         195.000000   
80   Adelie  Torgersen        34.60000       17.20000         189.000000   
81   Adelie  Torgersen        42.90000       17.60000         196.000000   
82   Adelie  Torgersen        36.70000       18.80000         187.000000   
83   Adelie  Torgersen        35.10000       19.40000         193.000000   
116  Adelie  Torgersen        38.60000       17.00000         188.000000   
117  Adelie  Torgersen        37.30000       20.50000         199.000000   
118  Adelie  Torgersen        35.70000       17.00000         189.000000   
119  Adelie  Torgersen        41.10000       18.60000         189.000000   
120  Adelie  Torgersen        36.20000       17.20000         187.000000   
121  Adelie  Torgersen        37.70000       19.80000         198.000000   
122  Adelie  Torgersen        40.20000       17.00000         176.000000   
123  Adelie  Torgersen        41.40000       18.50000         202.000000   
124  Adelie  Torgersen        35.20000       15.90000         186.000000   
125  Adelie  Torgersen        40.60000       19.00000         199.000000   
126  Adelie  Torgersen        38.80000       17.60000         191.000000   
127  Adelie  Torgersen        41.50000       18.30000         195.000000   
128  Adelie  Torgersen        39.00000       17.10000         191.000000   
129  Adelie  Torgersen        44.10000       18.00000         210.000000   
130  Adelie  Torgersen        38.50000       17.90000         190.000000   
131  Adelie  Torgersen        43.10000       19.2000

<h2>Insights:</h2>
<h3>The table shows Adelie penguins from Torgersen Island in 2009, filtered for those with flipper lengths over 200 mm. These individuals have body masses ranging from 3.0 kg to 4.3 kg, with males generally heavier than females. Bill lengths vary from 38.5 mm to 44.1 mm, and the longest flipper recorded is 210 mm.</h3>
<hr>

<h2>7. All Gentoo species penguins weighing less than 4000 g</h2>

In [28]:
dream_mid_flipper_penguins = penguins[(penguins['island'] == 'Dream') & (penguins['flipper_length_mm'] >= 190) & (penguins['flipper_length_mm'] <= 210)]
dream_mid_flipper_penguins

species island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
34      Adelie  Dream            36.4           17.0              195.0   
35      Adelie  Dream            39.2           21.1              196.0   
36      Adelie  Dream            38.8           20.0              190.0   
41      Adelie  Dream            40.8           18.4              195.0   
43      Adelie  Dream            44.1           19.7              196.0   
..         ...    ...             ...            ...                ...   
339  Chinstrap  Dream            55.8           19.8              207.0   
340  Chinstrap  Dream            43.5           18.1              202.0   
341  Chinstrap  Dream            49.6           18.2              193.0   
342  Chinstrap  Dream            50.8           19.0              210.0   
343  Chinstrap  Dream            50.2           18.7              198.0   

     body_mass_g     sex  year  mass_kg  
34        3325.0  female  2007      3.3  
35        4150.0    male  2007      4.2  
36        3950.0    male  2007      4.0  
41        3900.0    male  2007      3.9  
43        4400.0    male  2007      4.4  
..           ...     ...   ...      ...  
339       4000.0    male  2009      4.0  
340       3400.0  female  2009      3.4  
341       3775.0    male  2009      3.8  
342       4100.0    male  2009      4.1  
343       3775.0  female  2009      3.8  

[86 rows x 9 columns]

<h2>Insights:</h2>
<h3>The table shows penguins from Dream Island with flipper lengths between 190 mm and 210 mm, including Adelie and Chinstrap species. While the goal was to list Gentoo penguins under 4000 g</h3>
<hr>

<h2>8. What is the general average for each species in every island?</h2>

In [29]:
species_island = penguins.groupby(['species', 'island'])['body_mass_g'].mean()
species_island

species    island   
Adelie     Biscoe       3709.659091
           Dream        3688.392857
           Torgersen    3715.899123
Chinstrap  Dream        3733.088235
Gentoo     Biscoe       5068.965761
Name: body_mass_g, dtype: float64

<h2>Insight:</h2>
<h3>Gentoo penguins on Biscoe Island have the highest average body mass at about 5069 g, making them significantly heavier than other species in the dataset. In contrast, Adelie penguins have a lower average body mass across all three islands—Biscoe (≈ 3709 g), Dream (≈ 3688 g), and Torgersen (≈ 3716 g)—indicating a relatively consistent size regardless of location. Chinstrap penguins, found only on Dream Island in this dataset, have an average body mass of about 3733 g, which is slightly higher than the Adelie average on the same island.</h3>
<hr>

<h2>9. What is the sum of body mass per year?</h2>

In [30]:
penguins.groupby('year')['body_mass_g'].sum()

year
2007    453776.754386
2008    486400.000000
2009    505226.754386
Name: body_mass_g, dtype: float64

<h2>Insight:</h2>
<h3>The table presents the total body mass of all penguins recorded for each year in the dataset. In 2007, the cumulative body mass reached approximately 453,777 g, which then increased to 486,400 g in 2008. By 2009, the total body mass further rose to about 505,227 g.</h3>
<hr>

<h2>10. Give the yearly average mass for each species, island, and sex</h2>

In [31]:
penguins.groupby(['species', 'island', 'sex', 'year'])['mass_kg'].mean()

species    island     sex     year
Adelie     Biscoe     female  2007    3.480000
                              2008    3.244444
                              2009    3.450000
                      male    2007    3.800000
                              2008    4.011111
                              2009    4.275000
           Dream      female  2007    3.266667
                              2008    3.425000
                              2009    3.350000
                      male    2007    4.009091
                              2008    4.112500
                              2009    3.940000
           Torgersen  female  2007    3.450000
                              2008    3.525000
                              2009    3.187500
                      male    2007    3.991667
                              2008    4.187500
                              2009    3.787500
Chinstrap  Dream      female  2007    3.569231
                              2008    3.466667
                         

<h2>Insight:</h2>
<h3>Biscoe Island, male Adelie penguins in 2009 averaged 4.275 kg compared to females at 3.45 kg. Gentoo penguins, also on Biscoe, consistently have the highest average masses, with males reaching around 5.47 kg in 2007 and females about 4.79 kg in 2009.</h3>
<hr>

<h2>11. What is the average of body mass and flipper length in every species in island-sex-year?</h2>

In [32]:
penguins.groupby(['species', 'island', 'sex', 'year'])[['body_mass_g', 'flipper_length_mm']].mean()

body_mass_g  flipper_length_mm
species   island    sex    year                                
Adelie    Biscoe    female 2007  3470.000000         181.800000
                           2008  3244.444444         186.555556
                           2009  3446.875000         191.250000
                    male   2007  3770.000000         181.600000
                           2008  4011.111111         192.555556
                           2009  4268.750000         193.500000
          Dream     female 2007  3269.444444         185.000000
                           2008  3412.500000         189.000000
                           2009  3357.500000         189.500000
                    male   2007  4000.000000         187.727273
                           2008  4100.000000         195.000000
                           2009  3945.000000         192.800000
          Torgersen female 2007  3475.000000         187.625000
                           2008  3518.750000         190.000000
                           2009  3193.750000         187.250000
                    male   2007  3991.812865         191.326267
                           2008  4193.750000         193.500000
                           2009  3784.375000         198.625000
Chinstrap Dream     female 2007  3569.230769         188.692308
                           2008  3472.222222         192.666667
                           2009  3522.916667         194.333333
                    male   2007  3819.230769         196.153846
                           2008  4127.777778         202.777778
                           2009  3927.083333         201.833333
Gentoo    Biscoe    female 2007  4618.750000         211.062500
                           2008  4627.272727         213.000000
                           2009  4786.250000         213.700000
                    male   2007  5472.222222         218.722222
                           2008  5379.166667         221.750000
                           2009  5396.948099         221.621467

<h2>Insight:</h2>
<h3>Gentoo penguins on Biscoe Island stand out with the highest averages overall, with males in particular exceeding 5,300 g in body mass and over 218 mm in flipper length. In contrast, Adelie penguins show lower averages, especially females, whose body mass often stays below 3,500 g.</h3>
<hr>

<h2>12. What is the minimum and maximum flipper length per group</h2>

In [33]:
penguins.groupby(['species', 'island'])['flipper_length_mm'].agg(['min', 'max'])

min    max
species   island                      
Adelie    Biscoe     172.000000  203.0
          Dream      178.000000  208.0
          Torgersen  176.000000  210.0
Chinstrap Dream      178.000000  212.0
Gentoo    Biscoe     200.915205  231.0

<h2>Insight:</h2>
<h3>Gentoo penguins on Biscoe Island have the largest flippers overall, with lengths ranging from about 200.92 mm to 231 mm, highlighting their generally larger body size. In contrast, Adelie penguins on Biscoe Island have 
the shortest flippers, with a minimum of 172 mm. Chinstrap penguins, found only on Dream Island, have flipper lengths between 178 mm and 212 mm, placing them in the mid-to-upper range compared to other species.</h3>
<hr>

<h2>13. What is the average of bill length of all male Adelie in Dream island?</h2>

In [34]:
avg_bill_length = penguins[(penguins['species'] == 'Adelie') & (penguins['island'] == 'Dream') & (penguins['sex'] == 'male')]['bill_length_mm'].mean()
print(avg_bill_length)

39.98275862068966


<h2>Insight:</h2>
<h3>By filtering the table for rows where the species is "Adelie", the island is "Dream", and the sex is "male", it narrows the data to the exact group of interest. The calculation is then performed on the bill_length_mm column using the .mean() function, producing an average of approximately 39.98 mm.</h3>
<hr>

In [35]:
uk

Index Accident_Severity Accident_Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District_Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2       Single carriageway               Urban     Fine no high winds   
3       Single carriageway               Urban     Fine no high winds   
4       Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

<hr>

<h1>14. Compare number of accidents by Weather Condition</h1>

In [36]:
uk["Weather_Conditions"].value_counts()

Weather_Conditions
Fine no high winds       535013
Raining no high winds     79696
Other                     17150
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: count, dtype: int64

<h2>Insight:</h2>
<h3>By filtering the table for rows where the species is "Adelie", the island is "Dream", and the sex is "male", it narrows the data to the exact group of interest. The calculation is then performed on the bill_length_mm column using the .mean() function, producing an average of approximately 39.98 mm.</h3>
<hr>

<h1>15. What is the average number of casualties in Highland district</h1>

In [37]:
high = uk[uk['District_Area'] == 'Highland']['Number_of_Casualties'].mean()
print(high)

1.4750123701138051


<h2>Insight:</h2>
<h3>By filtering the table for rows where the species is "Adelie", the island is "Dream", and the sex is "male", it narrows the data to the exact group of interest. The calculation is then performed on the bill_length_mm column using the .mean() function, producing an average of approximately 39.98 mm.</h3>
<hr>

<h1>16. What is the average vehicles involved per road type</h1>

In [38]:
vehicles = uk.groupby(['Vehicle_Type', 'Road_Type']).size().reset_index(name='count')
print(vehicles)

                           Vehicle_Type           Road_Type  count
0                  Agricultural vehicle    Dual carriageway    293
1                  Agricultural vehicle      One way street     49
2                  Agricultural vehicle          Roundabout    129
3                  Agricultural vehicle  Single carriageway   1449
4                  Agricultural vehicle           Slip road     27
..                                  ...                 ...    ...
68  Van / Goods 3.5 tonnes mgw or under    Dual carriageway   5107
69  Van / Goods 3.5 tonnes mgw or under      One way street    723
70  Van / Goods 3.5 tonnes mgw or under          Roundabout   2294
71  Van / Goods 3.5 tonnes mgw or under  Single carriageway  25675
72  Van / Goods 3.5 tonnes mgw or under           Slip road    361

[73 rows x 3 columns]


<h2>Insight:</h2>
<h3>By filtering the table for rows where the species is "Adelie", the island is "Dream", and the sex is "male", it narrows the data to the exact group of interest. The calculation is then performed on the bill_length_mm column using the .mean() function, producing an average of approximately 39.98 mm.</h3>
<hr>

<h1>17. Show datas about serious accidents containing two casualties with two vehicles</h1>

In [41]:
casual_vehi = uk[(uk['Accident_Severity'] == 'Serious') & (uk['Number_of_Casualties'] > 2) & (uk['Number_of_Vehicles'] > 2)]
print(casual_vehi)

                Index Accident_Severity Accident_Date   Latitude  \
511     200701BS70539           Serious    14-10-2019  51.491861   
3335    200701EO40752           Serious    16-10-2019  51.529367   
4525    200701GD63185           Serious      2/6/2019  51.535459   
4747    200701GD66678           Serious      5/8/2019  51.570592   
4837    200701GD67253           Serious    30-09-2019  51.574618   
...               ...               ...           ...        ...   
660263  201063DP30510           Serious    19-09-2022  51.867575   
660310  201063DP35310           Serious    31-10-2022  52.686302   
660375  201063DP42110           Serious    23-11-2022  51.953352   
660394  201091NE00922           Serious    21-01-2022  57.597286   
660550  201091NJ08521           Serious    22-06-2022  58.307893   

              Light_Conditions           District_Area  Longitude  \
511      Darkness - lights lit  Kensington and Chelsea  -0.201432   
3335     Darkness - lights lit               

<h2>Insight:</h2>
<h3>By filtering the table for rows where the species is "Adelie", the island is "Dream", and the sex is "male", it narrows the data to the exact group of interest. The calculation is then performed on the bill_length_mm column using the .mean() function, producing an average of approximately 39.98 mm.</h3>
<hr>

In [50]:
uk

Index Accident_Severity Accident_Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District_Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2       Single carriageway               Urban     Fine no high winds   
3       Single carriageway               Urban     Fine no high winds   
4       Single carriageway               Urban     Fine no high winds   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

<h1>18. Show the accidents in daylight in Kensington and Chelsea with dry road</h1>

In [57]:
weather = uk[(uk['Light_Conditions'] == 'Daylight') & (uk['District_Area'] == 'Kensington and Chelsea') & (uk['Road_Surface_Conditions'] == 'Dry')][['Light_Conditions', 'District_Area', 'Road_Surface_Conditions']]
print(weather)

       Light_Conditions           District_Area Road_Surface_Conditions
3              Daylight  Kensington and Chelsea                     Dry
5              Daylight  Kensington and Chelsea                     Dry
6              Daylight  Kensington and Chelsea                     Dry
9              Daylight  Kensington and Chelsea                     Dry
11             Daylight  Kensington and Chelsea                     Dry
...                 ...                     ...                     ...
532789         Daylight  Kensington and Chelsea                     Dry
532823         Daylight  Kensington and Chelsea                     Dry
532829         Daylight  Kensington and Chelsea                     Dry
533039         Daylight  Kensington and Chelsea                     Dry
533316         Daylight  Kensington and Chelsea                     Dry

[1679 rows x 3 columns]


<h2>Insight:</h2>
<h3>By filtering the table for rows where the species is "Adelie", the island is "Dream", and the sex is "male", it narrows the data to the exact group of interest. The calculation is then performed on the bill_length_mm column using the .mean() function, producing an average of approximately 39.98 mm.</h3>
<hr>

<h1>19. Show the accidents happended in Rural area with between three to four number of vehicles only</h1>

In [73]:
urban = uk[(uk['Urban_or_Rural_Area'] == 'Urban') & (uk['Number_of_Vehicles'] >= 3) & (uk['Number_of_Vehicles'] <= 4)][['Urban_or_Rural_Area', 'Number_of_Vehicles']]
print(urban)

       Urban_or_Rural_Area  Number_of_Vehicles
2                    Urban                   3
3                    Urban                   4
5                    Urban                   3
19                   Urban                   4
37                   Urban                   3
...                    ...                 ...
660080               Urban                   3
660104               Urban                   3
660189               Urban                   4
660258               Urban                   3
660264               Urban                   3

[34776 rows x 2 columns]


<h2>Insight:</h2>
<h3>By filtering the table for rows where the species is "Adelie", the island is "Dream", and the sex is "male", it narrows the data to the exact group of interest. The calculation is then performed on the bill_length_mm column using the .mean() function, producing an average of approximately 39.98 mm.</h3>
<hr>